spccore/internal/lock.py

```diff
@@ -0,0 +1,114 @@
+DEFAULT_BLOCKING_TIMEOUT = datetime.timedelta(seconds=70)
+CACHE_UNLOCK_WAIT_TIME = 0.5
+
+
+class LockedException(Exception):
+    pass
+
+
+class Lock(object):
+    """
+    Implements a lock by making a directory named <lock_name>.lock
+    """
+    SUFFIX = 'lock'
+
+    def __init__(self,
+                 name: str,
+                 *,
+                 cwd: str = None,
+                 max_age: datetime.timedelta = LOCK_DEFAULT_MAX_AGE,
+                 default_blocking_timeout: datetime.timedelta = DEFAULT_BLOCKING_TIMEOUT
+                 ) -> None:
+        self.name = name
+        self.held = False
+        self.cwd = cwd if cwd else os.getcwd()
+        self.lock_dir_path = os.path.join(self.cwd, ".".join([name, Lock.SUFFIX]))
+        self.max_age = max_age
+        self.default_blocking_timeout = default_blocking_timeout
+
+    def blocking_acquire(self, *, timeout: datetime.timedelta = None, break_old_locks: bool = True) -> bool:
+        """
+        Acquire a lock.
+
+        :param timeout: the time frame to acquire the lock.
+            If none is set, this function will use the default_blocking_timeout.
+        :param break_old_locks: set to False to ignore old locks. Default True.
+        :return: True on success; otherwise throws a LockedException.
+        """
+        if self.held:
+            return True
+        if timeout is None:
+            timeout = self.default_blocking_timeout
+        lock_acquired = False
+        try_lock_start_time = time.time()
+        while time.time() - try_lock_start_time < timeout.total_seconds():
+            lock_acquired = self._acquire(break_old_locks=break_old_locks)
+            if lock_acquired:
+                break
+            else:
+                doze(CACHE_UNLOCK_WAIT_TIME)
+        if not lock_acquired:
+            raise LockedException("Could not obtain a lock on the file cache within timeout: {timeout}."
+                                  " Please try again later.".format(**{'timeout': str(timeout)}))
+
+    def release(self) -> None:
+        """Release lock or do nothing if lock is not held"""
+        if self.held:
+            try:
+                shutil.rmtree(self.lock_dir_path)
+                self.held = False
+            except OSError as err:
+                if err.errno != errno.ENOENT:
+                    raise
+
+    def _get_age(self) -> int:
+        """Return the age of the lock"""
+        try:
+            return time.time() - os.path.getmtime(self.lock_dir_path)
+        except OSError as err:
+            if err.errno != errno.ENOENT and err.errno != errno.EACCES:
+                raise
+            return 0
+
+    def _acquire(self, *, break_old_locks: bool = True) -> bool:
+        """
+        Attempt to acquire lock.
+
+        :param break_old_locks: set to False to ignore old locks. Default True.
+        :return: True on success; otherwise False.
+        """
+        if self.held:
```

Comment:
only entry point is blocking_acquire
remove redundant self.held check from the old code

Reply:
@zimingd , there are actually 2 public functions for Lock: blocking_acquire and release. As I understand, `self.held` is an optimization for both `blocking_acquire` and `release`. I'm not sure if it's redundant. Can you explain what else can we use instead of `self.held` for this optimization?